react/AppIcon/index.jsx

```diff
@@ -7,16 +7,129 @@ import Icon from '../Icon'
 
 import appDefaultIcon from '../../assets/icons/ui/cube.svg'
 
+let domain
+const _getDomain = () => {
+  if (domain) return domain
+  const root = document.querySelector('[role=application]')
+  if (!root) {
+    throw new Error('Cannot find [role=application] element')
+  }
+  domain = root.dataset.cozyDomain
+  if (!domain) {
+    throw new Error(
+      'Application element does not contain `data-cozy-domain` attribute'
+    )
+  }
+  return domain
+}
+
+let protocol
+const _getProtocol = () => {
+  if (protocol) return protocol
+  protocol = window.location.protocol
+  return protocol
+}
+
+const _preload = url => {
+  return new Promise((resolve, reject) => {
+    const loader = document.createElement('img')
+    loader.onload = () => {
+      resolve(url)
+    }
+    loader.onerror = () => {
+      reject(new Error(`Error while preloading ${url}`))
+    }
+    loader.src = url
+  })
+}
+
+const _getAppIconURL = app => {
+  let path = _getInstalledIconPath(app)
+  path = path || _getRegistryIconPath(app)
+  return path ? `${_getProtocol()}//${_getDomain()}${path}` : null
+}
+
+const _getInstalledIconPath = app => {
+  return app.links && app.links.icon
+}
+
+const _getRegistryIconPath = app => {
+  return (
+    app.latest_version &&
+    app.latest_version.version &&
+    `/registry/${app.slug}/${app.latest_version.version}/icon`
+  )
+}
+
+const _loaded = []
+
+const DONE = 'done'
+const ERRORED = 'errored'
+const FETCHING = 'fetching'
+const IDLE = 'idle'
+
 export class AppIcon extends Component {
   constructor(props, context) {
     super(props, context)
+
+    const { app } = props
+    let source
+    try {
+      source = _getAppIconURL(app)
+    } catch (error) {} // eslint-disable-line no-empty
+
+    const loaded = source && _loaded.includes(source)
```

Comment:
Hashes are a better fit to check for inclusions (no need to go through the whole array).

Reply:
I prefer hashes too, I don't even remember why I used `.includes`, exhaustion maybe. 👍